src/main/java/frc/robot/Robot.java

```diff
@@ -159,12 +193,117 @@ public void robotPeriodic() {
+            Commands.runOnce(
+                () -> {
+                  AutoAimStates.curShotSpeeds = speeds.get();
+                  Logger.recordOutput("AutoAim/cur shot speedd", AutoAimStates.curShotSpeeds);
+                  AutoAimStates.curShotData =
+                      AutoAim.shotMap.get(
+                          swerve
+                              .getLinearFuturePose(
+                                  AutoAim.LOOKAHEAD_TIME_SECONDS, AutoAimStates.curShotSpeeds)
+                              .minus(FieldConstants.getSpeaker())
+                              .getTranslation()
+                              .getNorm());
+                  System.out.println(Timer.getFPGATimestamp());
+                },
+                swerve),
+            Commands.deadline(
+                    Commands.waitSeconds(AutoAim.LOOKAHEAD_TIME_SECONDS),
+                    Commands.parallel(
+                        shooter.runStateCmd(
+                            () -> AutoAimStates.curShotData.getRotation(),
+                            () -> AutoAimStates.curShotData.getLeftRPM(),
+                            () -> AutoAimStates.curShotData.getRightRPM()),
+                        swerve.teleopPointTowardsTranslationCmd(
+                            () -> AutoAimStates.curShotSpeeds.vxMetersPerSecond,
+                            () -> AutoAimStates.curShotSpeeds.vyMetersPerSecond,
+                            AutoAim.LOOKAHEAD_TIME_SECONDS)))
+                .finallyDo(
+                    () -> {
+                      Logger.recordOutput("AutoAim/End Pose", swerve.getPose());
+                    }),
+            Commands.runOnce(
+                () -> {
+                  System.out.println("Shoot note");
+                }),
+            // keeps moving to prevent the robot from stopping and changing the velocity of the note
+            swerve
+                .runVelocityFieldRelative(
+                    () ->
+                        new ChassisSpeeds(
+                            AutoAimStates.curShotSpeeds.vxMetersPerSecond,
+                            AutoAimStates.curShotSpeeds.vyMetersPerSecond,
+                            0))
+                .withTimeout(0.25))
+        .andThen(() -> System.out.println(Timer.getFPGATimestamp()), swerve);
+  }
+
+  public Command autonomousAutoAim() {
+
+    return Commands.sequence(
+        Commands.deadline(
+                Commands.waitSeconds(AutoAim.LOOKAHEAD_TIME_SECONDS),
+                Commands.parallel(
+                    shooter.runStateCmd(
+                        () -> AutoAimStates.curShotData.getRotation(),
+                        () -> AutoAimStates.curShotData.getLeftRPM(),
+                        () -> AutoAimStates.curShotData.getRightRPM()),
+                    swerve.autonomousPointTowardsTranslationCmd()))
+            .andThen(
+                () -> {
+                  Logger.recordOutput("AutoAim/End Pose", swerve.getPose());
```

Comment:
This pattern seems like it happens a lot. You may want to considering creating a helper for this. Something like:

```java
// Assuming we make our own Logs class for helpers
LoggerCommands.recordOutput("AutoAim/EndPose", swerve::getPose) 
```

Reply:
The more compelling version of this is when you need multiples:

```java
LoggerCommands.recordOutputs("Foo", a::getFoo)
         .and("Bar", a::getBar)
         .andAlso("Baz", () -> moreComplicated(0));
```

Reply:
@kevinclark was wondering why to include an `andAlso` from `and`. Would there be a difference in semantics?

Reply:
If we're using a builder pattern at some point we need to actually build. So we could do multiple 'and' and then something to convert the builder to the built value. Or we could have a method that implies finality and does it. The latter just felt a little nicer if the builder is really just concatting. If there were other non-and operations I'd have a separate `build` method.